### ISSUE TITLE
UTF-8 matching fails on M1 MacOS with 5.4.9 

### ISSUE THREAD

**Issue by mlmitch** [VectorCamp/vectorscan#147]
Hello! First, thanks for the work you're doing on vectorscan. It's great.

As the title states, UTF-8 matching fails on M1 MacOS with 5.4.9.
I suspect this is caused by the fix to https://github.com/VectorCamp/vectorscan/issues/135 and I realize it might be solved through https://github.com/VectorCamp/vectorscan/issues/98.
Just want to make sure you're aware.
This issue is not present on M1 MacOS with 5.4.8.

The following hyperscan unit tests fail after compilation:
```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] Hyperscan/BadPattern.Block/145, where GetParam() = expr: "[\Q�\Eaaaa]", expected error: Expression is not valid UTF-8.
[  FAILED  ] Hyperscan/BadPattern.Stream/145, where GetParam() = expr: "[\Q�\Eaaaa]", expected error: Expression is not valid UTF-8.
```

Also, a test I run in my application fails that specifies `"星{2}"` with `HS_FLAG_UTF8` as the pattern and `"点星星灯"` is the test string. One match is expected but no match is produced.

Also, different sign comparison warnings are produced during compilation. For example:
```
.../Parser.cpp:857:13: warning: comparison of integers of different signs: 'const char' and 'unsigned int' [-Wsign-compare]
                if ( (*p) > 35u ) {
```

Finally, all these issues go away when supplying `-funsigned-char` to the build.

Some high-level details on my system:
M1 16-inch Macbook Pro
MacOS 12.6.5
AppleClang 14.0.0.14000029
Boost 1.82.0

**Comment by markos** [comment]
Could you please try current develop branch? We have MacOS X 13 in the CI now and it seems to be passing all the tests. Thanks.

**Comment by mlmitch** [comment]
I'm seeing the exact same failures as before.
I also see the same char sign warnings.

I got the develop source branch from https://github.com/VectorCamp/vectorscan/archive/refs/heads/develop.zip at about 9:20am Mountain Time today.

MacOS 13.6 now. Apple M1 Pro CPU. Apple Clang 15.0.0.15000040

Here are some redacted cmake and make logs for my project:
[cmake logs](https://gist.github.com/mlmitch/d67f75fcbf534b8781f5171d2941ba89)
[make logs](https://gist.github.com/mlmitch/b64205a804d28ec60e087df54612388b)

I can see the line in the cmake logs `RAGEL_C_FLAGS-Wno-unused -funsigned-char`
I'll just add that my `-funsigned-char` fix is specified with `add_compile_options(-funsigned-char)` in my project's top level `CMakeLists.txt` and it applies to the whole project.

**Comment by markos** [comment]
@mlmitch this is strange, could you please mention the full cmake command? No such warnings appear here (both for Debug and Release build types):

Compiler used: Apple clang version 14.0.3 (clang-1403.0.22.14.1)

cmake -DBUILD_STATIC_LIBS=True -DBUILD_SHARED_LIBS=True -DCMAKE_BUILD_TYPE={Debug,Release} -DFAT_RUNTIME=False -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ ../

I will try to check with a more recent clang. The problem is only with Ragel-generated code, so fixing RAGEL_C_FLAGS should be enough, but we have already done that.

**Comment by mlmitch** [comment]
If I remember correctly, the make warnings are new with Apple Clang 15.
The cmake command I use is `cmake -DCMAKE_BUILD_TYPE=Release`.
However, I set some relevant things in my CMakeLists.txt:
```
set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)

set(BUILD_SHARED_LIBS OFF CACHE BOOL "Don't build shared libraries for dependencies" FORCE)
set(BUILD_STATIC_LIBS ON CACHE BOOL "Build static libraries for dependencies" FORCE)

##### boost #####
cmake_policy(SET CMP0135 NEW)
FetchContent_Declare(BOOST
  URL https://boostorg.jfrog.io/artifactory/main/release/1.83.0/source/boost_1_83_0.tar.gz
  URL_HASH SHA512=231acce3021ef790ee0355e620faf375439c3f2ece0facac681acfc7be60a792c8404bd8c5920cd68c0d9dbcf8b8ac923afdd8d6e8f653306eef974b070e4490
)
FetchContent_MakeAvailable(BOOST)
set(BOOST_ROOT ${boost_SOURCE_DIR}) 

##### hyperscan/vectorscan #####
set(CMAKE_POLICY_DEFAULT_CMP0048 NEW)
set(CMAKE_POLICY_DEFAULT_CMP0115 OLD)
set(CMAKE_POLICY_DEFAULT_CMP0077 NEW)
cmake_policy(SET CMP0135 NEW)
FetchContent_Declare(HYPERSCAN
  URL  https://github.com/VectorCamp/vectorscan/archive/refs/heads/develop.zip
  URL_HASH SHA512=6943e689f97e8aa1644dff5a114864591ded8ef704f29f732c950c299e3e6e413b8d223c2ba4dbe9f62b21f0d3b39d456cb27bbedb1e2c7de8198216edbd74f7
)

# Fix for UTF-8 bug on ARM
# https://github.com/VectorCamp/vectorscan/issues/147
add_compile_options(-funsigned-char)

# vectorscan feature to prevent the export of hyperscan symbols in dynamic library
# need to define HS_PUBLIC_API as empty
# https://github.com/VectorCamp/vectorscan/issues/161
add_definitions(-DHS_PUBLIC_API=)

...
```

Hopefully I'm not offside here, but you are sure you are using and ARM processor? I realize you probably are, but it would explain the difference.

**Comment by markos** [comment]
@mlmitch we don't support MacOS X on Intel if that replies your question, our macos CI system is a mac mini M1. :)

**Comment by mlmitch** [comment]
Great. Just thought I'd make sure.

I can probably share a little more on a call if you want more details.

**Comment by markos** [comment]
@mlmitch I have just run the specific test with the UTF-8 chinese characters on the M1 with clang 14. It seems to pass the test. I have tried to do a search if Apple clang has changed default sign for char in clang>14 but I could not find anything relevant. Could you please verify the default sign for char? I don't object adding a special case for this, but I'd rather be 100% certain.

**Comment by mlmitch** [comment]
Looks to be signed.

```
% clang --version
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: arm64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

% clang main.c 

% ./a.out 
char is signed
```

main.c:

```
#include <stdio.h>
#include <limits.h>

int main() {
#if (CHAR_MIN != 0) 
   printf("char is signed\n");
#else 
   printf("char is unsigned\n");
#endif 

   return 0;
}
```

**Comment by markos** [comment]
I have updated my M1 to latest MacOS version and clang 1500.0.40.1.
I have tried building with default cmake invocation no arguments. The unit test totally passes on develop branch on both compilers (clang 14 and 15).

This is the unit test as written on the previous bug report:

```
#include "hs.h"

#include <string>
#include <iostream>
#include <cassert>

using namespace std;

int main() {

  std::string pattern = u8"星{2}";
  hs_compile_error_t *error = nullptr;
  hs_database_t *db         = nullptr;
  int flags = HS_FLAG_UTF8 | HS_FLAG_SINGLEMATCH;

  hs_error_t ret = hs_compile(pattern.c_str(), flags, HS_MODE_BLOCK,
                              nullptr,&db,&error);
  if (ret != HS_SUCCESS) {
      std::cout << "hs_compile error msg: " << error->message << ". expression: " << error->expression << std::endl;
      hs_free_compile_error(error);
      error = nullptr;
  }
  assert(ret == HS_SUCCESS);

  hs_scratch_t *scratch = nullptr;
  ret = hs_alloc_scratch(db, &scratch);
  assert(ret == HS_SUCCESS);

  std::string scanText = u8"星星点灯";
  static int matchCount = 0;
  ret = hs_scan(db, scanText.c_str(), scanText.size(), 0, scratch,
                [](unsigned int id, unsigned long long from,
                   unsigned long long to, unsigned int flags,
                   void *context) -> int {
        //printMatchedInfo(id, from, to, flags, context);
        matchCount++;
        return 0;
      },
      nullptr);

  assert(ret == HS_SUCCESS);
  std::cout << "matchCount = " << matchCount << std::endl;
  assert(matchCount == 1);

  // cleanup
  hs_free_database(db);
  hs_free_scratch(scratch);
  matchCount = 0;
}
```

I compiled this with the following command:

```
% clang++ -std=c++11 test.cpp -Ibuild-release/src -Ibuild-release -Isrc -Lbuild-release/lib -lhs -o test
% ./test
matchCount = 1
```

Without any extra `-funsigned-char` needed in the cmake. Please verify that it works for you, as I cannot reproduce this don't think it's a bug anymore.

I made certain I have also updated my brew environment to the latest versions.

**Comment by markos** [comment]
@mlmitch could you please confirm that this works for you without any further changes?

**Comment by mlmitch** [comment]
I was able to get vectorscan compiling on its own locally and all tests pass.
I wasn't able to get the above example compiling/linking properly today using the command line only like you did.

I did get it working in a mimimal cmake project that uses vectorscan as a dependency.
In this project, the above test produces a match count of zero.
The same unit tests also fail.
Here is that project:
https://github.com/mlmitch/vectorscan-test

It seems that the issue comes from using vectorscan through cmake?

Switching this project to vectorscan 5.4.8 fixes the problem.

**Comment by mlmitch** [comment]
I was now able to compile your example exactly as described above against `vectorscan-develop`.
It produces a match count of 1 as you expect.

**Comment by markos** [comment]
This is good to hear, so we can assume this bug is fixed now, thank  you for your feedback! :)

**Comment by mlmitch** [comment]
I think I figured out what is going wrong once we isolated the issue to vectorscan being brought in through CMake.
The change introduced in 5.4.9 applied unsigned char to ragel outputs using `set_source_files_properties`.
However, it seems vectorscan inherited some improper usage of `set_source_files_properties` from hyperscan.

The effect was that the unsigned char compile flag wasn't being applied to ragel output files.
The same is true of warning suppression flags which explains why I was seeing many compile warnings and you weren't.
I believe this is fixed with some small CMakeLists tweaks.

I have opened a pull request with them here:
https://github.com/VectorCamp/vectorscan/pull/189
